chalk-solve/src/recursive/solve.rs

```diff
@@ -68,12 +69,12 @@ pub(super) trait SolveIteration<I: Interner>: SolveDatabase<I> {
                 // or from the lowered program, which includes fallback
                 // clauses. We try each approach in turn:
 
-                let InEnvironment { environment, goal } = &canonical_goal.canonical.value;
-
+                let (_, _, InEnvironment { environment, goal }) =
+                    self.new_inference_table(&canonical_goal);
```

Comment:
What motivated this change?

Reply:
It was basically a result of noticing that in the SLG solver (which worked properly), but not the recursive solver, we instantiate goals before finding relevant program clauses, then wondering if replicating that might work here. It (surprisingly) did fix all the test failures, but it wasn't really driven by a strong intuition.